sphinx/ext/intersphinx/_load.py

```diff
@@ -138,36 +139,41 @@ def load_mappings(app: Sphinx) -> None:
     intersphinx_cache: dict[InventoryURI, InventoryCacheEntry] = inventories.cache
     intersphinx_mapping: IntersphinxMapping = app.config.intersphinx_mapping
 
+    expected_uris = {uri for _name, (uri, _invs) in intersphinx_mapping.values()}
+
+    for uri in frozenset(intersphinx_cache):
+        if intersphinx_cache[uri][0] not in intersphinx_mapping:
+            # Remove all cached entries that are no longer in `intersphinx_mapping`.
+            del intersphinx_cache[uri]
+        if uri not in expected_uris:
+            # Remove cached entries with a different target URI
+            # than the one in `intersphinx_mapping`.
+            # This happens when the URI in `intersphinx_mapping` is changed.
+            del intersphinx_cache[uri]
```

Comment:
@AA-Turner doesn't this create the possibility of a `KeyError` if both branches are followed? (or are we certain that they are mutually exclusive conditions?)

Reply:
(yes, it should be an elif)

Reply:
(I'm reasoning over it at the moment, but not certain either way yet)

Reply:
Adam's just split my 'or' actually so that's why it should be an 'elif'.

Reply:
I interpret the two conditions as:

* The cached `name` entry doesn't exist in the current Intersphinx mapping.
* The cached inventory URL doesn't exist in the Intersphinx mapping.

I think both of those could be true at the same time; for example when a mapping entry that was previously cached is removed.

Reply:
Good catch @jayaddison, thank you!